---
title: "Methylation data pre-processing"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
load("na.omit_TCGA_BRCA_methylation450_tsv_gz.Rdata") ## load the original data 
head(na.omit_TCGA_BRCA_methylation450_tsv_gz)

row.names(na.omit_TCGA_BRCA_methylation450_tsv_gz) <- na.omit_TCGA_BRCA_methylation450_tsv_gz[,1] ## removes the first column which is the cg i
head(na.omit_TCGA_BRCA_methylation450_tsv_gz)

na.omit_TCGA_BRCA_methylation450_tsv_gz <- na.omit_TCGA_BRCA_methylation450_tsv_gz[,-1] ## removes the first column which is the cg id

head(na.omit_TCGA_BRCA_methylation450_tsv_gz)

na.omit_TCGA_BRCA_methylation450.matrix <- as.matrix(na.omit_TCGA_BRCA_methylation450_tsv_gz) ## convert to matrix form

class(na.omit_TCGA_BRCA_methylation450.matrix) <- "numeric" ## converts the data to a numeric matrix 
head(na.omit_TCGA_BRCA_methylation450.matrix)
save(na.omit_TCGA_BRCA_methylation450.matrix,file = "na.omit_TCGA_BRCA_methylation450.matrix.Rdata") ## save the file 
```

######################## Filtering the data using champ #############################
# qsub -I -l select=1:ncpus=4:mem=60GB -l walltime=00:60:00 -A UQ-Health-SOM
cd $TMPDIR
pwd
/state/partition1/pbs/tmpdir/pbs.746630.awonmgr2
cd /home/uqmdeluc/
cp BRCA.fil.norm.batch.avail.Rdata $TMPDIR
cp samples.df.meth.Rdata $TMPDIR
cd /home/uqmdeluc/R-3.6.1/bin/
./R

```{r}
cd /home/uqmdeluc
library(ChAMP)
setwd("/state/partition1/pbs/tmpdir/pbs.746630.awonmgr2")
load("samples.df.meth.Rdata") # load the phenotype data
##generate a phenotype data
load("na.omit_TCGA_BRCA_methylation450.matrix.Rdata") # load the methylation data
##generate a phenotype data
dim(na.omit_TCGA_BRCA_methylation450.matrix) # 363791    890
samples.df.meth <- as.data.frame(colnames(na.omit_TCGA_BRCA_methylation450.matrix))
colnames(samples.df.meth) <- "Sample_Name"
samples.df.meth$Sample_Name <- as.character(samples.df.meth$Sample_Name)
samples.df.meth$Patient <- rapply(lapply(strsplit(samples.df.meth$Sample_Name,split = "-"),"[",1:3),f = paste,collapse = "-") ## add patient id only for matching purpose 
save(samples.df.meth,file = "samples.df.meth.Rdata")

# check if column of the matrix is the same as the phenotype data 
all(samples.df.meth$Sample_Name == colnames(na.omit_TCGA_BRCA_methylation450.matrix)) ## TRUE, therefore  proceed with further filtering 

BRCA.fil <- champ.filter(beta=na.omit_TCGA_BRCA_methylation450.matrix,
             M=NULL,
pd=samples.df.meth,
intensity=NULL,
Meth=NULL,
UnMeth=NULL,
detP=NULL,
beadcount=NULL,
autoimpute=F,
filterDetP=TRUE,
ProbeCutoff=0,
SampleCutoff=0.1,
detPcut=0.01,
filterBeads=TRUE,
beadCutoff=0.05,
filterNoCG = TRUE,
filterSNPs = TRUE,
population = NULL,
filterMultiHit = TRUE,
filterXY = TRUE,
fixOutlier = TRUE,
arraytype = "450K")
####
[===========================]
[<<<< ChAMP.FILTER START >>>>>]
-----------------------------

In New version ChAMP, champ.filter() function has been set to do filtering on the result of champ.import(). You can use champ.import() + champ.filter() to do Data Loading, or set "method" parameter in champ.load() as "ChAMP" to get the same effect.

This function is provided for user need to do filtering on some beta (or M) matrix, which contained most filtering system in champ.load except beadcount. User need to input beta matrix, pd file themselves. If you want to do filterintg on detP matrix and Bead Count, you also need to input a detected P matrix and Bead Count information.

Note that if you want to filter more data matrix, say beta, M, intensity... please make sure they have exactly the same rownames and colnames.


[ Section 1:  Check Input Start ]
  You have inputed beta for Analysis.

  pd file provided, checking if it's in accord with Data Matrix...
    pd file check success.

  Parameter filterDetP is TRUE, checking if detP in accord with Data Matrix...
    !!! Parameter detP is not found, filterDetP is reset FALSE now.

  Parameter filterBeads is TRUE, checking if beadcount in accord with Data Matrix...
    !!! Parameter beadcount is not found, filterBeads is reset FALSE now.

  Checking Finished :filterMultiHit,filterSNPs,filterNoCG,filterXY would be done on beta.
[ Section 1: Check Input Done ]


[ Section 2: Filtering Start >>

  Filtering NoCG Start
    Only Keep CpGs, removing 1024 probes from the analysis.

  Filtering SNPs Start
    Using general 450K SNP list for filtering.
    Filtering probes with SNPs as identified in Zhou's Nucleic Acids Research Paper 2016.
    Removing 21923 probes from the analysis.

  Filtering MultiHit Start
    Filtering probes that align to multiple locations as identified in Nordlund et al
    Removing 4 probes from the analysis.

  Filtering XY Start
    Filtering probes located on X,Y chromosome, removing 8272 probes from the analysis.

  Updating PD file
    filterDetP parameter is FALSE, so no Sample Would be removed.

  Fixing Outliers Start
    Replacing all value smaller/equal to 0 with smallest positive value.
    Replacing all value greater/equal to 1 with largest value below 1..
[ Section 2: Filtering Done ]

 All filterings are Done, now you have 332568 probes and 890 samples.

[<<<<< ChAMP.FILTER END >>>>>>]
[===========================]
[You may want to process champ.QC() next.]
 

############ Further filtering using the Zhou data##################

#Load the Zhou dataset 
HM450.hg38.manifest.tsv <- read.delim("HM450.hg38.manifest.tsv.gz", stringsAsFactors=FALSE)
##check if any prove in the general masked ce
probes.data <- row.names(BRCA.fil$beta)
length(probes.data)  # 332568
probes_masked <- HM450.hg38.manifest.tsv$probeID[HM450.hg38.manifest.tsv$MASK_general == T] ## identified the probes which should be masked 
# Identify if any probes that should be masked  but is not; this should be zero, hopefully
problem.probe <- probes.data[which(probes.data %in% probes_masked )]
length(problem.probe)
# proceed to remove these additional 2932 probes 
BRCA.fil.z <- BRCA.fil
BRCA.fil.z$beta <- BRCA.fil.z$beta[!(row.names(BRCA.fil.z$beta) %in% probes_masked),]
dim(BRCA.fil.z$beta) ## expect dim[1] == 329636 ??
save(BRCA.fil,file = "BRCA.fil.Rdata") ## save the initial file
save(BRCA.fil.z,file = "BRCA.fil.z.Rdata") ## save the initial file

########################## Data normalization #################################

BRCA.fil.norm <- champ.norm(beta=BRCA.fil.z$beta,
           resultsDir="./CHAMP_Normalization/",
           method="BMIQ",
           plotBMIQ=T,
           arraytype="450K",
           cores=4)
save(BRCA.fil.norm,file = "BRCA.fil.norm.Rdata")
```

########### Estimating the presence of batch effects in the methylation ################ 
# Batch effect data was obtained from https://bioinformatics.mdanderson.org/MQA/
```{r}
load("/home/uqmdeluc/methylation_preprocessing/BRCA.fil.norm.Rdata")
load("samples.df.met.batch.Rdata")
# filter the normalized data to isolate the samples for which we have batch infor 
BRCA.fil.norm.batch.avail <- BRCA.fil.norm[,which(colnames(BRCA.fil.norm) %in% samples.df.met.batch$Sample_Name)]
dim(BRCA.fil.norm.batch.avail) #  329636    884, as expected 
save(BRCA.fil.norm.batch.avail,file = "BRCA.fil.norm.batch.avail.Rdata")
## check order of samples in the pheno and the actual data 
all(colnames(BRCA.fil.norm.batch.avail) == samples.df.met.batch$Sample_Name)  ## all good ! 
# Estimate batch effects for TSS  
library(kBET)
input.kBET <- t(BRCA.fil.norm.batch.avail)
batch.estimate.TSS <- kBET(input.kBET,samples.df.met.batch$TSS)
save(batch.estimate.TSS,file = "batch.estimate.TSS.Rdata")
################### Estimate batch effects for batch_id ####################
# qsub -I -l select=1:ncpus=4:mem=60GB -l walltime=00:60:00 -A UQ-Health-SOM
cd $TMPDIR
pwd
/state/partition1/pbs/tmpdir/pbs.747131.awonmgr2
cd /home/uqmdeluc/
cp BRCA.fil.norm.batch.avail.Rdata $TMPDIR
cp samples.df.meth.Rdata $TMPDIR
cd /home/uqmdeluc/R-3.6.1/bin/
./R
```{r}
setwd("/state/partition1/pbs/tmpdir/pbs.747131.awonmgr2")
load("BRCA.fil.norm.batch.avail.Rdata")
load("/home/uqmdeluc/methylation_preprocessing/samples.df.met.batch.Rdata")
## check order of samples in the pheno and the actual data 
all(colnames(BRCA.fil.norm.batch.avail) == samples.df.met.batch$Sample_Name)  ## all good ! 

library(kBET)
input.kBET <- t(BRCA.fil.norm.batch.avail)
batch.estimate.batch_id <- kBET(input.kBET,samples.df.met.batch$batch_id)
save(batch.estimate.batch_id,file = "batch.estimate.batch_id.Rdata")
```
## assess the presence of batch effects in the dats 
```{r}
batches <- read.delim("~/Downloads/e6665e30546434303dd06d1c296a30ac-data/original/batches.tsv") # downloaded data from https://bioinformatics.mdanderson.org/MQA/
# identify which of these samples are tumopur samples  
batches$tumour_status <- lapply(strsplit(x = batches$Sample,split = "-"),"[[",4)
batches$tumour_status <- as.numeric(substr(batches$tumour_status ,1,nchar(batches$tumour_status )-1))
batches.tum <- batches[which(batches$tumour_status %in% seq(1-9)),]


samples.df.met.batch <- samples.df.meth[which(samples.df.meth$Patient %in% batches.tum$Patient),]# 884/890 available
dim(samples.df.met.batch) # 884   2884   2
setwd("/Users/xavierdeluca/Desktop/post_doc/")
save(samples.df.met.batch,file = "samples.df.met.batch.Rdata") ##sample information for those samples with batch information 


#  Because they collapse samples for the same patient, batch information becomes problematic to source.....
batches.tum.avail <- batches.tum[which(batches.tum$Patient %in% samples.df.meth$Patient),] ## only keep batch info for the sample we need 
dim(batches.tum.avail) ## 794  11 represents 794 unique patients
## fill in the batch effect info : 
samples.df.met.batch$TSS <- NA ## add TSS info
samples.df.met.batch$ship_date <- NA ## cant add this one as theres multiple netries for each patient for some of the samples...
 samples.df.met.batch$batch_id <- NA # add batch id info
for(i in seq_along(1:dim(samples.df.met.batch)[1])){
  samples.df.met.batch$TSS[i] <- unique(batches.tum$TSS[which(batches.tum$Patient == samples.df.met.batch$Patient[i])])
  #samples.df.met.batch$ship_date[i] <- list(unique(batches.tum$ShipDate[which(batches.tum$Patient == samples.df.met.batch$Patient[i])]))
  samples.df.met.batch$batch_id[i] <-  unique(batches.tum$BatchId[which(batches.tum$Patient == samples.df.met.batch$Patient[i])])

}

 save(samples.df.met.batch,file = "samples.df.met.batch.Rdata") ## save ethe samples for which we have methylation and batch info ready to go.

batches.tum.avail.test <- unique(batches.tum.avail[,c("Sample","Patient","BatchId","TSS")])
```
## decided not to further correct for batch effeects given the findings by MBatch online ! 
Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

