---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
batches.RNA <- read_delim("~/Downloads/e6665e30546434303dd06d1c296a30ac-data 2/original/batches.tsv", 
     delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE)
batches.RNA$Sample_Name <- rapply(f = paste, object = lapply(strsplit(x = batches.RNA$Sample, split = "-"),"[",1:4),collapse = "-")
# Check the number of unique samples 
length(unique(batches.RNA$Sample_Name)) # 1217 unique samples, therefore some duplication present 
test <- batches.RNA[which(batches.RNA$Sample_Name %in% batches.RNA$Sample_Name[which(duplicated(batches.RNA$Sample_Name) == T)]),]
```
# read in the RNAseq data obtained from https://xenabrowser.net/datapages/?dataset=TCGA-BRCA.htseq_counts.tsv&host=https%3A%2F%2Fgdc.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443 selected  the RNAseq-HTSeq counts
```{r}
TCGA.BRCA.htseq_counts.tsv.gz <- read.delim("~/Desktop/post_doc/TCGA-BRCA.htseq_counts.tsv.gz.tsv")
samples.RNA.df <- as.data.frame(colnames(TCGA.BRCA.htseq_counts.tsv.gz)[-1])
colnames(samples.RNA.df) <- "Sample_Name"
samples.RNA.df$Sample_Name <- gsub(pattern = "\\.",replacement = "-",x = samples.RNA.df$Sample_Name)
samples.RNA.df$Patient <- rapply(f = paste,object = lapply(strsplit(split = "-", samples.RNA.df$Sample_Name),"[",1:3),collapse = "-")
# check which one has batch data 
length(which(samples.RNA.df$Patient %in% batches.RNA$Patient)) # 1216 samples  (includes normals and tumours)
samples.RNA.df$Tumour_status <- lapply(strsplit(samples.RNA.df$Sample_Name,split = "-"),"[[",4) ## obtain the tumour status of the tumour
samples.RNA.df$Tumour_status <- substr(samples.RNA.df$Tumour_status,1,nchar(samples.RNA.df$Tumour_status)-1) ## remove the last letter
## rename  the dataframe 
colnames(TCGA.BRCA.htseq_counts.tsv.gz)  <-  gsub(pattern = "\\.",replacement = "-",x = colnames(TCGA.BRCA.htseq_counts.tsv.gz))
row.names(TCGA.BRCA.htseq_counts.tsv.gz) <- TCGA.BRCA.htseq_counts.tsv.gz$Ensembl_ID
TCGA.BRCA.htseq_counts.tsv.gz$Ensembl_ID <- NULL

TCGA.BRCA.htseq_counts.fil <- TCGA.BRCA.htseq_counts.tsv.gz[,which(colnames(TCGA.BRCA.htseq_counts.tsv.gz) %in% batches.RNA$Sample_Name)]
dim(TCGA.BRCA.htseq_counts.fil)  # 60488  1217  ## sample batch info available for all samples! 
save(TCGA.BRCA.htseq_counts.fil,file = "TCGA.BRCA.htseq_counts.fil.Rdata")
head(TCGA.BRCA.htseq_counts.fil[1:10,1:10])

```
# need to further remove the samples for which  there was aggregation of the data in a single point of methylation.
```{r}
TCGA.BRCA.htseq_counts.fil.no_agg <- TCGA.BRCA.htseq_counts.fil[,-which(colnames(TCGA.BRCA.htseq_counts.fil) %in% batches.RNA$Sample_Name[which(duplicated(batches.RNA$Sample_Name) == T)])]
dim(TCGA.BRCA.htseq_counts.fil.no_agg) #  60488  1212 lost 5 samples

```
# Load the annotation data for the gene
```{r}
 gencode_v22_annotation_gene <- read_delim("~/Downloads/gencode.v22.annotation.gene.probeMap","\t", escape_double = FALSE, trim_ws = TRUE)

```
## Deciding on the pathway for preprocessing the RNAseq data 
```{r}
library(DESeq2)
## need to round to integer  # https://support.bioconductor.org/p/105964/
## remove the previous tranformation applied to the data (https://www.biostars.org/p/383669/)
TCGA.BRCA.htseq_counts.fil.no_agg.no_log <- 2^TCGA.BRCA.htseq_counts.fil.no_agg  
head(TCGA.BRCA.htseq_counts.fil.no_agg.no_log[1:5,1:5]) ##  log2(count+1) removed the log2 
TCGA.BRCA.htseq_counts.fil.no_agg.no_log <- TCGA.BRCA.htseq_counts.fil.no_agg.no_log - 1 ## remove the  + 1
head(TCGA.BRCA.htseq_counts.fil.no_agg.no_log[1:5,1:5]) ##  log2(count+1) removed the log2 
TCGA.BRCA.htseq_counts.fil.no_agg.rounded <- round(TCGA.BRCA.htseq_counts.fil.no_agg.no_log)
head(TCGA.BRCA.htseq_counts.fil.no_agg.rounded[1;5,1:5]) ## rounded 
# generating the phenotype dataframe for the  dataframe 
batches.RNA.fil.no_agg.rounded <- batches.RNA[which(batches.RNA$Sample_Name %in% colnames(TCGA.BRCA.htseq_counts.fil.no_agg.rounded)),]
batches.RNA.fil.no_agg.rounded <- batches.RNA.fil.no_agg.rounded[match(colnames(TCGA.BRCA.htseq_counts.fil.no_agg.rounded),batches.RNA.fil.no_agg.rounded$Sample_Name),]
#check the order of the names to make sure it matches
all(colnames(TCGA.BRCA.htseq_counts.fil.no_agg.rounded) == batches.RNA.fil.no_agg.rounded$Sample_Name) # TRUE
row.names(batches.RNA.fil.no_agg.rounded) <- batches.RNA.fil.no_agg.rounded$Sample_Name
# creagte a DESeq object 
head(TCGA.BRCA.htseq_counts.fil.no_agg.rounded[1:10,1:10])
DEseq.RNA.no_agg<- DESeqDataSetFromMatrix(countData = TCGA.BRCA.htseq_counts.fil.no_agg.rounded,
                                    colData = batches.RNA.fil.no_agg.rounded,
                                    design = ~ 1)

DEseq.RNA.no_agg.dds <- estimateSizeFactors(DEseq.RNA.no_agg)  # Accounts for the different sequencing depths among your libraries 

idx <- rowSums(counts(DEseq.RNA.no_agg.dds) >= 5 ) >= 3 ## filtering 
DEseq.RNA.no_agg.dds.fil <- DEseq.RNA.no_agg.dds[idx,]
save(DEseq.RNA.no_agg.dds.fil,file = "DEseq.RNA.no_agg.dds.fil.Rdata")


```
# qsub -I -l select=1:ncpus=1:mem=50GB -l walltime=00:60:00 -A UQ-Health-SOM
```{r}
library(DESeq2)
setwd("/state/partition1/pbs/tmpdir/pbs.749191.awonmgr2")
load("DEseq.RNA.no_agg.dds.fil.Rdata")
## running the DESeq function on the cluster  + VST transformation
DEseq.RNA.no_agg.dds.fil.run <- DESeq(DEseq.RNA.no_agg.dds.fil)  ## run the DESeq algorithm
save(DEseq.RNA.no_agg.dds.fil.run,file = "DEseq.RNA.no_agg.dds.fil.run.Rdata")
vst.TCGA <- vst(DEseq.RNA.no_agg.dds.fil.run, blind=FALSE)  # VST tranformation to take out the varian from the data of interest !
save(vst.TCGA,file  = "vst.TCGA.Rdata")
```
## Estimate the possible presence of batch effects in the data
```{r}
setwd("/state/partition1/pbs/tmpdir/pbs.749193.awonmgr2")
load("vst.TCGA.Rdata")
load("batches.RNA.fil.no_agg.rounded.Rdata")
input.rna <- assay(vst.TCGA)
head(input.rna[1:10,1:10])
## check order of samples in the pheno and the actual data 
all(colnames(input.rna) == batches.RNA.fil.no_agg.rounded$Sample_Name)  ## all good ! 

library(kBET)
input.kBET <- t(input.rna)
batch.estimate.batch_id.rna <- kBET(input.kBET,batches.RNA.fil.no_agg.rounded$BatchId)
save(batch.estimate.batch_id.rna,file = "batch.estimate.batch_id.rna.Rdata")

batch.estimate.TSS.rna <- kBET(input.kBET,batches.RNA.fil.no_agg.rounded$TSS)
save(batch.estimate.TSS.rna,file = "batch.estimate.TSS.rna.Rdata")

batch.estimate.ShipDate.rna <- kBET(input.kBET,batches.RNA.fil.no_agg.rounded$ShipDate)
save(batch.estimate.ShipDate.rna,file = "batch.estimate.ShipDate.rna")

pca.data.RNA <- prcomp(input.kBET, center=TRUE) #compute PCA representation of the data
save(pca.data.RNA,file = "pca.data.RNA.Rdata")
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

